onnxscript/function_libs/tools/torch_lib/deduce_type_constraints.py

```diff
@@ -0,0 +1,403 @@
+    onnx.AttributeProto.INTS: "tensor(int64)",
+}
+
+
+class TypeConstraint:
+    """Type constraint shared by multiple values."""
+
+    name: str
+    type_strs: Set[str]
+    values: Set[Value]
+
+    def __init__(self, name: str, type_strs: Set[str]):
+        self.name = name
+        self.type_strs = type_strs
+        self.values = set()
+
+    def merge_type_constraint(self, other: TypeConstraint):
+        """Merge two type constraints.
+
+        Typically this process tightens the type constraints.
+
+        For example, consider two values with different existing type constraints:
+
+            input1: TypeConstraint(name="input1", type_strs={"tensor(float)", "tensor(int64)"})
+            input2: TypeConstraint(name="input2", type_strs={"tensor(float)", "tensor(int32)"})
+
+        Now consider they are two inputs to `Add` node, where the op schema is:
+
+            inputs: [T, T]
+            outputs: [T]
+            T: ['tensor(uint8)', ..., 'tensor(float)', 'tensor(int64)', 'tensor(int32)']
+
+        `input1` and `input2` must now bound to the same type constraint, hence the result is
+        the intersection of the two existing type constraints with the op schema type constraint:
+
+            input1: TypeConstraint(name="input1.input2", type_strs={"tensor(float)"})
+            input2: TypeConstraint(name="input1.input2", type_strs={"tensor(float)"})
+        """
+        if self is other:
+            return
+        self.type_strs.intersection_update(other.type_strs)
+        self.values = self.values.union(other.values)
+        for meta in self.values:
+            meta.type_constraint = self
+        self.name = f"{self.name}.{other.name}"
+
+    def bind_value(self, value: Value):
+        """Bind a value to this type constraint.
+
+        If the value does not have a type constraint, set it to this type constraint.
+        Otherwise, merge the two type constraints.
+        """
+        if value.type_constraint is not None:
+            self.merge_type_constraint(value.type_constraint)
+        else:
+            self.values.add(value)
+            value.type_constraint = self
+
+    def __repr__(self):
+        value_names = [value.name for value in self.values]
+        return f"TypeConstraint(name={self.name}, type_strs={self.type_strs}, values={value_names})"
+
+
+class Value:
+    """Represents a value in the graph. Associated with a TypeConstraint."""
+
+    def __init__(self, name: str):
+        self.type_constraint: Optional[TypeConstraint] = None
+        self.name: str = name
+
+    def merge_type_constraint(self, other: Value):
+        if other.type_constraint is not None:
+            other.type_constraint.bind_value(self)
+        elif self.type_constraint is not None:
+            self.type_constraint.bind_value(other)
+        else:
+            raise ValueError(
+                f"Cannot merge two values without type constraints. {self} and {other}"
```

Comment:
It could be done if None is treated as the set of all types. This may be useful to eliminate the restriction that inputs must have the "TensorType" annotation. Of course, the set of all types is infinite, but in practice, ONNX too restricts to types of depth 2 or 3 (like sequences of tensors, etc.). So, two choices to do this would be to use some depth-bounded set of all typestrs, or use None as a shorthand for that in class TypeConstraint also.